dftimewolf/lib/collectors/grr_base.py

```diff
@@ -0,0 +1,79 @@
+"""Base GRR module class. GRR modules should extend it."""
+
+import tempfile
+import time
+
+from grr_api_client import api as grr_api
+from grr_api_client import errors as grr_errors
+
+from dftimewolf.lib.module import BaseModule
+
+# This class does not implement process() since it is a base class.
+class GRRBaseModule(BaseModule):  # pylint: disable=abstract-method
+  """Base module for GRR hunt and flow modules.
+
+  Attributes:
+    output_path: Path to store collected artifacts.
+    grr_api: GRR HTTP API client.
+    reason: Justification for GRR access.
+    approvers: list of GRR approval recipients.
+  """
+  _CHECK_APPROVAL_INTERVAL_SEC = 10
+
+  def __init__(self, state):
+    super(GRRBaseModule, self).__init__(state)
+    self.reason = None
+    self.grr_api = None
+    self.approvers = None
+    self.output_path = None
+
+  # pylint: disable=arguments-differ
+  def setup(self, reason, grr_server_url, grr_auth, approvers=None):
+    """Initializes a GRR hunt results collector.
+
+    Args:
+      reason: justification for GRR access.
+      grr_server_url: GRR server URL.
+      grr_auth: Tuple containing a (username, password) combination.
+      approvers: list of GRR approval recipients.
+    """
+
+    self.approvers = []
+    if approvers:
+      self.approvers = [item.strip() for item in approvers.strip().split(',')]
+    self.grr_api = grr_api.InitHttp(api_endpoint=grr_server_url, auth=grr_auth)
+    self.output_path = tempfile.mkdtemp()
+    self.reason = reason
+
+  def _check_approval_wrapper(self, grr_object, function):
+    """Wraps a call to GRR functions checking for approval."""
+    approval_sent = False
+
+    while True:
```

Comment:
I suggest adding a TTL here, and raise if necessary

Reply:
By TTL you mean a max duration / number of tries? It's tricky because it can take from 30 seconds to 2h to get an approval granted...

Reply:
Ah, yes I remember now. Cool, fine to leave as is.